### PR TITLE
[7293] Changed mail microservice to use boost::process instead of system() (main)

### PR DESCRIPTION
**Pull request by FifthPotato**
Tested by hand to check functionality is as expected, but haven't run test suites yet.
A few questions:
I deleted the block with the solaris_platform stuff; is that okay? ...do we even support Solaris?
I also deleted the block for empty subjects, since in my testing, an empty subject still sent properly, but with no subject line.
Finally, not sure if checkStringForSystem() is the best name anymore for that function. I did a recursive grep and couldn't find anything else using it, but maybe we need to keep it for other peoples' clients? Should I write a new function with a better name and leave it as-is?

### CODE REVIEW

File: lib/core/src/stringOpr.cpp [hunk 1]
```diff
@@ -316,7 +316,7 @@ checkStringForSystem( const char * inString ) {
     if ( inString == NULL ) {
         return 0;
     }
-    if ( boost::regex_match( inString, boost::regex( "[a-zA-Z0-9,./ ]*" ) ) ) {
+    if (boost::regex_match(inString, boost::regex("[^\r\n]*"))) {
```
Comment (trel) [inline lib/core/src/stringOpr.cpp:319]: let's get a comment to say in words what this regex is doing (and why)... to be kind for the next reader/debugger.

File: server/re/src/mailMS.cpp [hunk 2]
```diff
@@ -87,20 +88,11 @@ int msiSendMail( msParam_t* xtoAddr, msParam_t* xsubjectLine, msParam_t* xbody,
         return SYS_MALLOC_ERR;
     }
 
-#ifdef solaris_platform
```
Comment (trel) [inline server/re/src/mailMS.cpp:90]: removing solaris is fine, yes.

File: server/re/src/mailMS.cpp [hunk 3]
```diff
@@ -87,20 +88,11 @@ int msiSendMail( msParam_t* xtoAddr, msParam_t* xsubjectLine, msParam_t* xbody,
...
-        sprintf( mailStr, "cat %s| mail -s '%s'  '%s'", fName, subjectLine, toAddr );
-    }
-    else {
-        sprintf( mailStr, "cat %s| mail  '%s'", fName, toAddr );
-    }
-#endif
     int ret = 0;
-    ret = system( mailStr );
+    namespace bp = boost::process;
+    ret = bp::system(bp::search_path("mail"), "-s", subjectLine, toAddr, bp::std_in < fName);
```
Comment (trel) [inline server/re/src/mailMS.cpp:93]: can combine to `int ret = ...` ?

**Review by trel — COMMENTED** [review]
and did we figure out how to test this?   throw some edge cases at it to confirm it behaves?  thinking about when we expand to other operating systems again, we want to have bumpers to know where we fall over / hit the corners.

File: lib/core/src/stringOpr.cpp (hunk 1 above)
Comment (korydraughn) [reply lib/core/src/stringOpr.cpp:319]: Now that we're using boost process, is this check still necessary?

File: server/re/src/mailMS.cpp [hunk 4]
```diff
@@ -87,20 +88,11 @@ int msiSendMail( msParam_t* xtoAddr, msParam_t* xsubjectLine, msParam_t* xbody,
...
     if ( ret ) {
-        irods::log( ERROR( ret, "mailStr command returned non-zero status" ) );
+        irods::log(ERROR(ret, "boost::process::system command returned non-zero status"));
```
Comment (korydraughn) [inline server/re/src/mailMS.cpp:95]: Boost process is an implementation detail and doesn't provide enough information to help the admin know where to start looking.

How do you feel about something like ... `msiSendMail error: <error_code>`?

That immediately tells the admin it's specific to this MSI.

Comment (korydraughn) [inline server/re/src/mailMS.cpp:95]: Let's use the new logger instead of `irods::log`.

File: server/re/src/mailMS.cpp [hunk 5]
```diff
@@ -1,10 +1,11 @@
-/// \file
+// \file
```
Comment (korydraughn) [inline server/re/src/mailMS.cpp:1]: Need that third forward slash for doxygen. Without it, doxygen may not produce documentation for this file.

**Review by korydraughn — COMMENTED** [review]
Good stuff.

File: lib/core/src/stringOpr.cpp (hunk 1 above)
Comment (FifthPotato) [reply lib/core/src/stringOpr.cpp:319]: Ah, so this check is actually for spec reasons: subject lines aren't allowed to include \r or \n by spec...thing is, in my tests, `mail` handles it just fine (by simply not adding a subject line)...but I thought it'd be good to still check.

Comment (trel) [reply lib/core/src/stringOpr.cpp:319]: yep - leave it and a comment saying the 'why'.

**Comment by korydraughn** [comment]
There should be information on which tests failed. How many tests are failing?

**Comment by korydraughn** [comment]
Also, please mark any review comments you've handled as resolved.

**Comment by korydraughn** [comment]
Good detective work.

I see a couple options for resolving this.
- (a) Declare a package dependency for installing mail
- (b) Detect whether the mail binary is installed and respond accordingly _(your current solution)_
- (c) We replace the use of the mail binary with libcurl's SMTP support

I think we have to do (a) and (b).

Thoughts?

**Comment by FifthPotato** [comment]
> Good detective work.
> 
> I see a couple options for resolving this.
> 
>     * (a) Declare a package dependency for installing mail
> 
>     * (b) Detect whether the mail binary is installed and respond accordingly _(your current solution)_
> 
>     * (c) We replace the use of the mail binary with libcurl's SMTP support
> 
> 
> I think we have to do (a) and (b).
> 
> Thoughts?

Since we agreed that maybe adding more deps is bad, I'll see how painful it is to do (c) and maybe we can go from there.

**Comment by alanking** [comment]
@FifthPotato - What's the status of this PR?

**Comment by FifthPotato** [comment]
> @FifthPotato - What's the status of this PR?

Dormant? Currently focusing on some of the other tickets.

**Comment by korydraughn** [comment]
This isn't high priority right now. Let's focus on other issues first. We'll come back to this later.

**Comment by korydraughn** [comment]
`msiSendMail` is now deprecated.

This PR can be closed.
